code/datums/components/traps.dm

```diff
@@ -0,0 +1,77 @@
+//for having one object make another do something when the first receives a signal
+
+#define COMSIG_TRAP_ACTIVATE "activate_trap" //sent to traps to activate them
+#define COMSIG_TRAP_LINK "link_trap" //a trap sent this will store ref of parent of component sending it in linked_datums
+#define COMSIG_TRAP_UNLINK "unlink_trap" //a trap sent this will remove ref of parent of component sending it in linked_datums
+
+/datum/component/trap
+	var/list/linked_datums = list() //holds references to other traps, triggers use it to activate linked traps
+	var/datum/callback/obj_activation_callback //callback invoked on activation, for easy handling of special effects that need properties/methods of parent
+
+	//todo - various commonly needed trap features
+	var/list/allowed_parent_types = list()
+	var/activation_sound
+	var/starts_invisible
+	var/del_on_activation
+
+/datum/component/trap/Initialize(_datums_to_link, _obj_activation_callback)
+	for(var/datum/target in _datums_to_link)
+		linked_datums += target
+		SEND_SIGNAL(target, COMSIG_TRAP_LINK)
+	obj_activation_callback = _obj_activation_callback
+	RegisterSignal(parent, COMSIG_TRAP_LINK, .proc/trap_link)
+	RegisterSignal(parent, COMSIG_TRAP_UNLINK, .proc/trap_unlink)
+	RegisterSignal(parent, COMSIG_TRAP_ACTIVATE, .proc/activate) 
+
+/datum/component/trap/proc/activate()
+	//override with custom behavior
+	obj_activation_callback?.Invoke()
+	if(isatom(parent))
+		var/atom/A = parent
+		A.visible_message("The [A] activates!") //debug message, todo: remove
+	//todo: playsound, etc
+
+/datum/component/trap/proc/trap_link(datum/component/source)
+	linked_datums += source.parent
+	
+/datum/component/trap/proc/trap_unlink(datum/component/source)
+	linked_datums -= source.parent
```

Comment:
Instead of doing this, you could add all the linked_datums into the same list that is shared amongst all components.

And instead of that again, you don't actually have to add a seperate component to all the linked_datums. One would absolutely be enough, but in that case I'm not sure that a component would actually be the best way to go about it. This is more akin to something like a ductnets or powernet datum than a component